fms/modules/embedding.py

```diff
@@ -1,21 +1,129 @@
 
 from fms.distributed.tensorparallel import (
     all_gather_from_tensor_model_parallel_region,
     apply_colwise_tp,
     apply_embedding_tp,
-    apply_rowwise_tp,
     copy_to_tensor_model_parallel_region,
-    reduce_from_tensor_model_parallel_region,
 )
 
 
+class AbsolutePositionEmbedding(nn.Module):
+    """Special form of embedding that includes a position encoding"""
+
+    def __init__(
+        self,
+        vocab_size: int,
+        emb_dim: int,
+        padding_idx: Optional[int] = None,
+        max_pos: int = 512,
+    ):
+        """
+        Initialize an AbsolutePositionEmbedding
+
+        Parameters
+        ----------
+        vocab_size: int
+            the length of the vocabulary
+        emb_dim: int
+            dimensionality of latent space
+        padding_idx: int, optional
+            Padding token index in the vocabulary. Typically, the token for
+            which positions will be zeroed out. If negative or None, no positions
+            will be zeroed. (default is None)
+        max_pos: int
+            the maximum possible sequence length (default is 512)
+        """
+        super().__init__()
+        self.max_pos = max_pos
+        self.emb_dim = emb_dim
+
+        # if negative padding_idx is given, just set to None
+        if padding_idx is not None:
+            padding_idx = (
+                padding_idx if padding_idx >= 0 and padding_idx < vocab_size else None
+            )
+        self.padding_idx = padding_idx
+
+        if self.padding_idx is None:
+            self.emb = nn.Embedding(vocab_size, self.emb_dim)
+        else:
+            self.emb = nn.Embedding(
+                vocab_size, self.emb_dim, padding_idx=self.padding_idx
+            )
+
+        self.pos_emb = nn.Embedding(max_pos, self.emb_dim)
+        self.register_buffer("pos_id", torch.arange(max_pos).unsqueeze(0))
+
+    def reset_params(self):
+        # Defaults to norm-preserving in reverse op, unit vector in forward op
+        layers = ["emb"]
+        layers.append("pos_emb")
+        for layer in layers:
+            nn.init.trunc_normal_(
+                getattr(self, layer).weight, mean=0.0, std=self.emb_dim**-0.5
+            )
+        # Preserve pad index dummy-hood
+        if self.padding_idx is not None:
+            self.emb.weight.data[self.padding_idx].zero_()
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        correct_pads: bool = False,
+    ):
+        """
+        perform a forward pass of absolute positional embedding
+
+        Parameters
+        ----------
+        x: torch.Tensor
+            the input tensor
+        position_ids: torch.LongTensor, optional
+            a tensor which signifies the positions in the position embedding. If
+            None, will simply be a range from beginning to end of sequence for
+            each sequence in the batch (default is None)
+        correct_pads: bool
```

Comment:
Maybe I need to think about it more but I'm questioning if this is the best way to handle padding. Having a user pass position_ids and/or flags like this makes the api pretty complicated (relative to just using nn.Embeddings directly).

Reply:
Problem is, a lot of times the position_ids are based on things like the pads or external factors such as the kv-cache. We did not want to include that in here as it seems out of scope. Also, it is possible the user has provided their own correction to the position_ids, but if not, it seems somewhat non-trivial to require the user to correct their own position_ids (it was an option though that was explored)

Reply:
> We did not want to include that in here as it seems out of scope.

I'm not sure what you mean here. I'm saying the scope of this class seems large in a way that doesn't seem to encapsulate behavior well.

when I said "user" I meant "user of this class," from an API perspective.

The user from a generation-perspective is presumably either not passing pos_ids or is going to be passing a single position_id for the next token, when using kv_caching.  they shouldn't need to worry about padding.

If I need to handle absolute position embeddings in a model I could do something like:
```
pos_ids = position_ids if position_ids is not None else (x != pad_idx).cumsum(1)
pos = self.position_embedding( pos_ids )
```
Switching to this class it seems like I need to read the code to understand it, and there are three parameters to forward, and I'm not really sure that it's making the model simpler.

Reply:
The above is essentially being handled in the AbsolutePositionEmbedding's forward (as it includes word embedding + position encoder), and the parameters are optional, by default if you just provide the position_ids, self.position_embedding will be called on those position_ids. The idea of putting the word embedding + position together was to simplify the logic for the user so they could call a single:

```python
self.emb(x) # optionally providing position_ids if they have got them already
```

rather than the user just hold both a 

```python
self.emb = ...
self.pos_emb = ...

x_emb = self.emb(x)
# do all necessary logic with x to get the position_ids how you want it
out = self.pos_emb(position_ids) + x_emb
```

`correct_pads` is a utility so it does not fall on the user to do this. By default we don't correct for padding, we could just never correct for padding here and just require the user to do that. This really was meant to handle the logic so the underlying class was not required to do so.

Reply:
ok for now will add it there for now, if we see we need to re-use later, we can address then